pkg/k8s/client.go

```diff
@@ -35,82 +37,94 @@ const (
 
 // KubeInfo holds the data required to talk to a cluster
 type KubeInfo struct {
-	Kubeconfig   string
-	context      string
-	namespace    string
-	clientConfig clientcmd.ClientConfig
+	flags       *pflag.FlagSet
+	configFlags *genericclioptions.ConfigFlags
+	config      *rest.Config
+	namespace   string
 }
 
 // NewKubeInfo returns a useable KubeInfo, handling optional
 // kubeconfig, context, and namespace.
 func NewKubeInfo(configfile, context, namespace string) *KubeInfo {
-	// Find the correct kube config file
-	configfilesearch := clientcmd.NewDefaultClientConfigLoadingRules()
-	if len(configfile) != 0 {
-		configfilesearch.ExplicitPath = configfile
-	}
+	flags := pflag.NewFlagSet("KubeInfo", pflag.PanicOnError)
+	result := NewKubeInfoFromFlags(flags)
 
-	// Possibly override context and namespace
-	overrides := &clientcmd.ConfigOverrides{}
-	if len(context) != 0 {
-		overrides.CurrentContext = context
+	var args []string
+	if configfile != "" {
+		args = append(args, "--kubeconfig", configfile)
 	}
-	if len(namespace) != 0 {
-		overrides.Context.Namespace = namespace
+	if context != "" {
+		args = append(args, "--context", context)
+	}
+	if namespace != "" {
+		args = append(args, "--namespace", namespace)
 	}
 
-	// Construct the config
-	kubeconfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(configfilesearch, overrides)
-
-	res := KubeInfo{
-		configfile,
-		context,
-		namespace,
-		kubeconfig,
+	err := flags.Parse(args)
+	if err != nil {
+		panic(err)
 	}
+	return result
+}
 
-	return &res
+// NewKubeInfoFromFlags adds the generic kubeconfig flags to the
+// provided FlagSet, and returns a *KubeInfo that configures itself
+// based on those flags.
+func NewKubeInfoFromFlags(flags *pflag.FlagSet) *KubeInfo {
+	configFlags := genericclioptions.NewConfigFlags(false)
+
+	// We can disable or enable flags by setting them to
+	// nil/non-nil prior to calling .AddFlags().
+	//
+	// .Username and .Password are already disabled by default in
+	// genericclioptions.NewConfigFlags().
+	//
+	// Unlike client-go/discovery.CachedDiscoveryClient,
+	// ericchiang/k8s doesn't support caching to the filesystem,
+	// so disable the '--cache-dir' flag.
+	configFlags.CacheDir = nil
```

Comment:
I think we can drop `configFlags.CacheDir`?  We at least intend to support caching, there's a TODO comment about it.  And we probably don't need to mention `ericchiang/k8s` :)